docs/design/2022-03-16-functional-dependency.md

```diff
@@ -0,0 +1,147 @@
+
+## Motivation And Background
+
+Since we have already known the dependencies between sets of attributes (columns) after the FD has been built, we can take use of this relationship to do optimization in many scenarios. Eg: in only-full-group-by mode, for non-aggregated columns in the select list, as long as it is functionally dependent on the group-by clause columns, then we can consider it legal. Otherwise, to extract the constant, equivalence, dependency relationship between columns everytime we use it is tricky and costly.  
+
+After trials and errors of only-full-group-by check, TiDB ultimately seek to functional dependency to solve it's only-full-group-by check problem completely. Besides, functional dependency can also be applied to many other scenarios to reduce complexity, for example, distinct elimination with/without order by, selectivity for the correlated column and so on.
+
+## Detail Design
+
+### Implementation Overview
+
+* How to store function dependency
+* How to maintain function dependency
+* How to integrate function dependency with TiDB arch
+
+#### How to store function dependency
+
+Generally speaking, functional dependencies essentially store the relationship between columns. According to paper [CS-2000-11.thesis.pdf](https://cs.uwaterloo.ca/research/tr/2000/11/CS-2000-11.thesis.pdf), this relationship can be constant, equivalence, strict and lax. Except constant is a unary, all others can be regarded as binary relationships. Therefore, we can use graph theory to store functional dependencies, maintaining edge sets between point-set and point-set, and attaching equivalence, strict and lax attributes as a label on each edge. For unary constant property, we can still save it as an edge, and since constant has no starting point-set, let's regard it as a headless edge.
+
+* equivalence: {a} == {b}
+* strict FD:   {a} -> {b}
+* lax FD:      {a} ~> {b}
+* constant:    { } -> {b}
+
+For computation convenience, we can optimize equivalence as {a,b} == {a,b}, so we don't have to search for the both direction side when confirming whether a point is in the equivalence set.
+
+#### How to maintain function dependency
+
+For open-sourced mysql, the construction and maintenance of FD is still relatively heavy. It uses sub-queries as logical units and uses nested methods to recursively advance FD. For each logical unit, the source for where FD is derived is relatively simple, basically from three elements there: group-by, where-clause and datasource(join/table). However, according to the paper [CS-2000-11.thesis.pdf](https://cs.uwaterloo.ca/research/tr/2000/11/CS-2000-11.thesis.pdf), TiDB chooses to build FDs based on logical operators as an alternative, without the boundary restrictions of logical sub-queries. 
+
+Example:
+
+```sql
+create table t1(a int, c int)
+create table t2(a int, b int)
+select t1.a from t1 join t2 on t2.a=t1.a group by t2.a where c=1;
```

Comment:
What's the purpose of this SQL? It's different from the below one.

Reply:
Ahaaa, a mistake, >.<